src/Generator/MethodGenerator.php

```diff
@@ -378,12 +379,30 @@ private static function extractReturnTypeFromMethodReflection(MethodReflection $
             return null;
         }
 
-        $returnTypeString = (string) $returnType;
+        if (! method_exists($returnType, 'getName')) {
+            return self::expandLiteralType((string) $returnType, $methodReflection);
+        }
+
+        return ($returnType->allowsNull() ? '?' : '')
+            . self::expandLiteralType($returnType->getName(), $methodReflection);
+    }
 
-        if ('self' === strtolower($returnType)) {
+    /**
+     * @param string           $literalReturnType
+     * @param ReflectionMethod $methodReflection
+     *
+     * @return string
+     */
+    private static function expandLiteralType($literalReturnType, ReflectionMethod $methodReflection)
```

Comment:
> If so, should it also consider static and $this? (If it's strictly from declared return type, ignore.)

`static` is not a valid type hint (for good reasons). `static` is used when resolving methods to call inside an instance, but it is not possible to use it as a hint due to its natural non-LSP-compliance.

`$this` is not a type

In general, reflecting from docblocks is now a deprecated feature (there are explicit tests that verify that docblocks aren't used for reflecting anymore)